---
title: "Discipline Project 2 - Varied Number of Genes - GSE120396" 
author: "Cailin Feldman"
date: "11/05/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r,include=FALSE}
library(R.utils)
library(reshape2)
library(ggplot2)
library(limma)
library(dplyr)
library(GEOquery)
library(Biobase)
library(biomaRt)
library(tidyverse)
```

```{r,include=FALSE}
# #get biocLite()
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.11")
```
```{r,include=FALSE}
BiocManager::install(c("GenomicFeatures", "AnnotationDbi", "GEOquery", "affy", "limma", "gcRMA"))
```
Opening GSE120396 From GEO

```{r, warning FALSE, error = FALSE, message = FALSE}
RNA_seq_GSE120396 <-getGEO("GSE120396")

RNA_seq_GSE120396<- RNA_seq_GSE120396$GSE120396_series_matrix.txt.gz

print(RNA_seq_GSE120396$characteristics_ch1.1[1:10])

RNA_seq_GSE120396_rejection_status  <- RNA_seq_GSE120396$characteristics_ch1.1
RNA_seq_GSE120396_rejection_status <- unlist(lapply( strsplit(as.character(RNA_seq_GSE120396_rejection_status), ": " ) , `[[` , 2)  )
table(RNA_seq_GSE120396_rejection_status)
```

Opening GSE120396 From Folder

```{r}
datadir = "GSE120396_RAW/"

# Read in the files
fileNames <- list.files(datadir)

# Check that we have read in the correct files
print(fileNames[1:5])
```

```{r, eval = FALSE, include=FALSE}
# unzip all the files
#for (files in fileNames){
#  gunzip(file.path(datadir,files))
#}
```

```{r}
# Read in all files to make a table
GSE120396 = c()
for(i in 1:length(fileNames)){
  temptable <- read.delim(file.path(datadir, fileNames[i]), header=TRUE)
  GSE120396 <- cbind(GSE120396, temptable[,2])
  colnames(GSE120396)[i] <- colnames(temptable)[2]
}

#Reading in the GSE
rownames(GSE120396) = read.delim(file.path(datadir, fileNames[1]), header=TRUE)[,1]

```

```{r, include= FALSE}
#Checks the dimensions It has 11721 rows and 88 columns, meaning for each of the 88 patients, there is gene expression measurement across 11721 genes.
dim(GSE120396)

#We need to check if our dataset has been log2 transformed, as this is a standard normalization technique for RNA-seq data. If not, then we need to log2 our data.
summary(GSE120396[,1])
```

```{r,  include= FALSE}
write.csv(GSE120396, "GSE120396_expression_matrix.txt")
```

```{r}
t_GSE120396 <- t(GSE120396_expression_matrix)
```

Principal Component Analysis (PCA) 

```{r}
GSE120396_pca <- prcomp(t(GSE120396))
GSE120396_df_toplot <- data.frame(RNA_seq_GSE120396_rejection_status, 
                        pc1 = GSE120396_pca$x[,1], pc2 = GSE120396_pca$x[,2]  )


g <- ggplot(GSE120396_df_toplot, aes(x = pc1, y = pc2, color = RNA_seq_GSE120396_rejection_status)) + 
  geom_point() + 
  theme_minimal() 
g
```

Modeling with GSE120396 (Proving The Choice Of SVM)

```{r}
largevar = apply(GSE120396, 1, var)
ind = which(largevar > quantile(largevar, 0.9))

X = as.matrix(t(GSE120396[ind,]))
y = RNA_seq_GSE120396_rejection_status
```

```{r}
cvK = 5  # number of CV folds
cv_50acc5_knn = cv_50acc5_svm = cv_50acc5_rf = c()
cv_acc_knn = cv_acc_svm = cv_acc_rf = c()

n_sim = 25 ## number of repeats
for (i in 1:n_sim) {

  cvSets = cvTools::cvFolds(nrow(X), cvK)  # permute all the data, into 5 folds
  cv_acc_knn = cv_acc_svm = cv_acc_rf = c()
  
  for (j in 1:cvK) {
    test_id = cvSets$subsets[cvSets$which == j]
    X_test = X[test_id, ]
    X_train = X[-test_id, ]
    y_test = y[test_id]
    y_train = y[-test_id]
    
    ## KNN
    fit5 = class::knn(train = X_train, test = X_test, cl = y_train, k = 5)
    cv_acc_knn[j] = table(fit5, y_test) %>% diag %>% sum %>% `/`(length(y_test))
    
    ## SVM
    svm_res <- e1071::svm(x = X_train, y = as.factor(y_train))
    fit <- predict(svm_res, X_test)
    cv_acc_svm[j] = table(fit, y_test) %>% diag %>% sum %>% `/`(length(y_test))

    ## RandomForest
    rf_res <- randomForest::randomForest(x = X_train, y = as.factor(y_train))
    fit <- predict(rf_res, X_test)
    cv_acc_rf[j] = table(fit, y_test) %>% diag %>% sum %>% `/`(length(y_test))
  }
  cv_50acc5_knn <- append(cv_50acc5_knn, mean(cv_acc_knn))
  cv_50acc5_svm <- append(cv_50acc5_svm, mean(cv_acc_svm))
  cv_50acc5_rf <- append(cv_50acc5_rf, mean(cv_acc_rf))
} ## end for
```

```{r}
boxplot(list(SVM = cv_50acc5_svm, KNN = cv_50acc5_knn , RF= cv_50acc5_rf ))
```

Modeling with GSE120396 (With Manipulated Number of Expressed Genes) (With Cross Validation)

```{r}
t_GSE120396[1,][21:41]
#GSE120396[,1][20:40]
```

```{r}
reduced_GSE120396 = GSE120396[20:40,]
largevar = apply(reduced_GSE120396, 1, var)
ind = which(largevar > quantile(largevar, 0.9))

X = as.matrix(t(reduced_GSE120396[ind,]))
y = RNA_seq_GSE120396_rejection_status
```

```{r}
cvK = 5  # number of CV folds
cv_50acc5_knn = cv_50acc5_svm = cv_50acc5_rf = c()
cv_acc_knn = cv_acc_svm = cv_acc_rf = c()

n_sim = 25 ## number of repeats
for (i in 1:n_sim){

  cvSets = cvTools::cvFolds(nrow(X), cvK)  # permute all the data, into 5 folds
  cv_acc_knn = cv_acc_svm = cv_acc_rf = c()
  
  for (j in 1:cvK){
    test_id = cvSets$subsets[cvSets$which == j]
    X_test = X[test_id, ]
    X_train = X[-test_id, ]
    y_test = y[test_id]
    y_train = y[-test_id]
    
    #print(X_train)
    #print(y_train)
    
     ## SVM
    svm_res <- e1071::svm(x = X_train, y = as.factor(y_train))
    fit <- predict(svm_res, X_test)
    cv_acc_svm[j] = table(fit, y_test) %>% diag %>% sum %>% `/`(length(y_test))
  }
  cv_50acc5_svm <- append(cv_50acc5_svm, mean(cv_acc_svm))
}
```

```{r}
#cv_50acc5_svm
boxplot(cv_50acc5_svm)
summary(cv_50acc5_svm)
```

Modeling with GSE120396 (With Manipulated Number of Expressed Genes) (Without Cross Validation)

```{r}
t_GSE120396[1,][21:41]
#GSE120396[,1][20:40]
```

```{r}
reduced_GSE120396 = GSE120396[20:40,]
largevar = apply(reduced_GSE120396, 1, var)
ind = which(largevar > quantile(largevar, 0.9))

ran <- sample(1:nrow(reduced_GSE120396), 0.9 * nrow(reduced_GSE120396)) 

X = as.matrix(t(reduced_GSE120396[ind,]))
reduced_y = RNA_seq_GSE120396_rejection_status
```

```{r}
ran <- sample(1:nrow(X), 0.9 * nrow(X)) 
 
X_test = X[ran, ]
X_train = X[-ran, ]
y_test = reduced_y[ran]
y_train = reduced_y[-ran]

## SVM
svm_res <- e1071::svm(x = X_train, y = as.factor(y_train))
fit <- predict(svm_res, X_test)
acc_svm = table(fit, y_test) %>% diag %>% sum %>% `/`(length(y_test))
acc_svm
```
Selecting The Best Genes 

```{r}
genes.2 <- read.csv("~/Documents/New DATA3888/genes-2.csv", header=FALSE)
# Just First Column with All rows
genes.2_resp = genes.2[, 1]
```

```{r}
short <- genes.2[,20:100]  
short_resp = genes.2_resp[:100]
model <- glm(short_resp ~., data = short, family = binomial) %>% stepAIC(trace = FALSE)
summary(model) 
```

You cant look at each indivual gene, you need to look at groups hence why a decion tree is the best 
